modules/icgc-argo-workflows/checkinput/resources/usr/bin/dnaaln.py

```diff
@@ -0,0 +1,377 @@
+            raise AssertionError("'single_end' should be specifed as \"True\" or \"False\".") 
+    
+
+    def _validate_read_group_count(self, row):
+        """Assert that expected read_group_count is correct."""
+        if len(row[self._read_group_count_col]) <= 0:
+            raise AssertionError("'read_group_count' input is required.")
+    
+    def _validate_experiment(self, row):
+        """Assert that expected Experiment is correct."""
+        if len(row[self._experiment_col]) <= 0:
+            raise AssertionError("'experiment' input is required.")
+        for val in ["WGS","WXS","RNA-Seq","Bisulfite-Seq","ChIP-Seq","Targeted-Seq"]:
+            if val==row[self._experiment_col]:
+                return
+        raise AssertionError("'experiment' type does not match the following: \"WGS\",\"WXS\",\"RNA-Seq\",\"Bisulfite-Seq\",\"ChIP-Seq\",\"Targeted-Seq\".")
+
+    def _validate_analysis_json(self, row):
+        """Assert that expected analysis_json is correct."""
+        if len(row[self._analysis_json_col]) <= 0:
+            raise AssertionError("'analysis_json' input is required.")
+        if not row[self._analysis_json_col].endswith(".json"):
+            raise AssertionError("'analysis_json' input should have the suffix \".json\".")
+
+    def validate_unique_samples(self):
+        """
+        Assert that the combination of sample name and FASTQ filename is unique.
+
+        In addition to the validation, also rename all samples to have a suffix of _T{n}, where n is the
+        number of times the same sample exist, but with different FASTQ files, e.g., multiple runs per experiment.
+
+        """
+        if len(self._seen) != len(self.modified):
+            raise AssertionError("The pair of sample name and FASTQ must be unique.")
+        seen = Counter()
+        for row in self.modified:
+            sample = row[self._sample_col]
+            seen[sample] += 1
+
+
+def read_head(handle, num_lines=10):
+    """Read the specified number of lines from the current position in the file."""
+    lines = []
+    for idx, line in enumerate(handle):
+        if idx == num_lines:
+            break
+        lines.append(line)
+    return "".join(lines)
+
+
+def sniff_format(handle):
+    """
+    Detect the tabular format.
+
+    Args:
+        handle (text file): A handle to a `text file`_ object. The read position is
+        expected to be at the beginning (index 0).
+
+    Returns:
+        csv.Dialect: The detected tabular format.
+
+    .. _text file:
+        https://docs.python.org/3/glossary.html#term-text-file
+
+    """
+    peek = read_head(handle)
+    handle.seek(0)
+    sniffer = csv.Sniffer()
+    dialect = sniffer.sniff(peek)
+    return dialect
+
+
+def check_samplesheet(file_in, file_out):
+    required_columns = {"patient","sex","status","sample","lane","fastq_1","fastq_2","single_end","experiment"}
+    # See https://docs.python.org/3.9/library/csv.html#id3 to read up on `newline=""`.
```

Comment:
For alignment workflow, we will need to gather information to generate `read_group` for aligner to use.
- "sample","lane","fastq_1","fastq_2","single_end": required fields
- "patient","sex","status","experiment": optional fields with default values if not provided
- "library_name", "platform_unit", "platform" should also be required fields so as to build `read_group`

Reply:
read_group info is enforced and capture in:
https://github.com/icgc-argo-workflows/argo-modules/pull/25#discussion_r1412578890